api/src/main/java/org/apache/iceberg/io/LocationProvider.java

```diff
@@ -37,6 +38,17 @@ public interface LocationProvider extends Serializable {
    */
   String newDataLocation(String filename);
 
+  /**
+   * Return a fully-qualified data file location for the given filename and options.
+   *
+   * @param filename a file name
+   * @param options options for deciding the location
+   * @return a fully-qualified location URI for a data file
+   */
+  default String newDataLocation(String filename, Map<String, String> options) {
```

Comment:
Where are these options coming from?

Reply:
These options is provided when user wants to write data to data file directly and need a valid location, the values would depend on the specific implementation - the unit tests shows a simple example.

Reply:
It still isn't clear to me how these new options are passed through an engine like Spark or Trino. I don't think that we can add this until we understand where this data is coming from.

Reply:
It's not particularly designed for well integrated engines, for the sample case in unit test, data writer may want to created data files in local region and add to table with iceberg api like `AppendFiles.appendFile`, with extra options, we can avoid users hard-coding locations by implementing a custom LocationProvider.

Reply:
I don't think we can merge this unless we have a clear idea of where the data is coming from. Otherwise, I think it makes more sense to implement your own location provider. There needs to be a use case around how the extra metadata gets to this new method.

Reply:
We can workaround this if this is not generic enough for OSS version, closing this PR, thanks for looking at this.